### PR TITLE
Fixed #31277 -- Relaxed system check of m2m intermediary tables for db_table collision when database routers are installed.

### PR DESCRIPTION
**Pull request by Qu4tro**
https://code.djangoproject.com/ticket/31277

### CODE REVIEW

File: django/db/models/fields/related.py [hunk 1]
```diff
@@ -1436,12 +1436,25 @@ def _get_field_name(model):
                 clashing_obj = '%s.%s' % (opts.label, _get_field_name(model))
             else:
                 clashing_obj = model._meta.label
+
```
Comment (felixxm) [inline django/db/models/fields/related.py]: Chop blank line.

File: django/db/models/fields/related.py [hunk 2]
```diff
@@ -1436,12 +1436,25 @@ def _get_field_name(model):
...
+            if settings.DATABASE_ROUTERS:
+                error_class, error_id = checks.Warning, 'fields.W344'
+                error_hint = (
+                    'You have configured settings.DATABASE_ROUTERS. '
+                    "Verify that '%s' and '%s' are correctly routed to separate databases."
```
Comment (felixxm) [inline django/db/models/fields/related.py]: Please wrap at 79 chars.

File: django/db/models/fields/related.py [hunk 3]
```diff
@@ -1436,12 +1436,25 @@ def _get_field_name(model):
...
+            clashing_pair = (m2m_db_table, clashing_obj)
             return [
-                checks.Error(
+                error_class(
                     "The field's intermediary table '%s' clashes with the "
-                    "table name of '%s'." % (m2m_db_table, clashing_obj),
+                    "table name of '%s'." % clashing_pair,
                     obj=self,
-                    id='fields.E340',
+                    hint=(error_hint % clashing_pair) if error_hint else None,
```
Comment (felixxm) [inline django/db/models/fields/related.py]: Using `clashing_pair` in a hint is misleading. We should use appropriate model names not field names or table names.

File: docs/releases/3.0.4.txt [hunk 4]
```diff
@@ -20,3 +20,7 @@ Bugfixes
   related fields or parent link fields with :ref:`multi-table-inheritance` in
   the ``of`` argument, the corresponding models were not locked
   (:ticket:`31246`).
+
+* Relaxed the system check added in Django 1.9.8 for models to reallow use of
+  the same ``db_table`` by multiple intermediary tables when database routers
+  are installed (:ticket:`31277`).
```
Comment (felixxm) [inline docs/releases/3.0.4.txt]: This patch doesn't qualify for a backport. Please remove release notes.

File: django/db/models/fields/related.py [hunk 5]
```diff
@@ -1436,12 +1436,25 @@ def _get_field_name(model):
...
+                error_class, error_id = checks.Error, 'fields.E340'
+                error_hint = None
+
+            clashing_pair = (m2m_db_table, clashing_obj)
             return [
-                checks.Error(
+                error_class(
                     "The field's intermediary table '%s' clashes with the "
-                    "table name of '%s'." % (m2m_db_table, clashing_obj),
+                    "table name of '%s'." % clashing_pair,
```
Comment (felixxm) [inline django/db/models/fields/related.py]: Revert this change.

File: django/db/models/fields/related.py [hunk 6]
```diff
@@ -1436,12 +1436,25 @@ def _get_field_name(model):
...
+            if settings.DATABASE_ROUTERS:
+                error_class, error_id = checks.Warning, 'fields.W344'
+                error_hint = (
+                    'You have configured settings.DATABASE_ROUTERS. '
+                    "Verify that '%s' and '%s' are correctly routed to separate databases."
+                )
+            else:
+                error_class, error_id = checks.Error, 'fields.E340'
+                error_hint = None
+
```
Comment (felixxm) [inline django/db/models/fields/related.py]: Chop blank line.

Comment (Qu4tro) [reply django/db/models/fields/related.py]: Resolved.

File: django/db/models/fields/related.py (hunk 5 above)
Comment (Qu4tro) [reply django/db/models/fields/related.py]: Resolved.

File: docs/releases/3.0.4.txt (hunk 4 above)
Comment (Qu4tro) [reply docs/releases/3.0.4.txt]: Resolved.

File: django/db/models/fields/related.py (hunk 2 above)
Comment (Qu4tro) [reply django/db/models/fields/related.py]: Resolved.

File: django/db/models/fields/related.py (hunk 1 above)
Comment (Qu4tro) [reply django/db/models/fields/related.py]: Resolved.

File: django/db/models/fields/related.py (hunk 3 above)
Comment (Qu4tro) [reply django/db/models/fields/related.py]: I might be in the wrong here or misunderstand, but isn't it possible to have an implicit model, backed by an actual table, just by having a `ManyToMany` field? I'm not sure which model name to use here.
Although, I do agree that the way `error_hint` is currently phrased, it's misleading. 

Do you have any suggestion for the required changes?
In the meantime, I've replaced `clashing_pair` by it's definition, on this instance.

**Comment by Qu4tro** [comment]
Glad to help :D And thanks for the review.

Resolved most of the notes. The `clashing_pair` one is still unresolved, because I'm not sure on how to handle it. 

Feedback on how to proceed would be great.

File: django/db/models/fields/related.py (hunk 3 above)
Comment (felixxm) [reply django/db/models/fields/related.py]: This is a field error, so IMO it's fine to mention only clashing model in the hint, e.g.
```diff
diff --git a/django/db/models/fields/related.py b/django/db/models/fields/related.py
index a2e939fa3d..b4babe30ac 100644
--- a/django/db/models/fields/related.py
+++ b/django/db/models/fields/related.py
@@ -1433,15 +1433,16 @@ class ManyToManyField(RelatedField):
                         if field.remote_field.through is model:
                             return field.name
                 opts = model._meta.auto_created._meta
-                clashing_obj = '%s.%s' % (opts.label, _get_field_name(model))
+                clashing_model = opts.label
+                clashing_obj = '%s.%s' % (clashing_model, _get_field_name(model))
             else:
-                clashing_obj = model._meta.label
+                clashing_model = model._meta.label
+                clashing_obj = clashing_model
             if settings.DATABASE_ROUTERS:
                 error_class, error_id = checks.Warning, 'fields.W344'
                 error_hint = (
-                    'You have configured settings.DATABASE_ROUTERS. '
-                    "Verify that '%s' and '%s' are correctly "
-                    'routed to separate databases.'
+                    'You have configured settings.DATABASE_ROUTERS. Verify '
+                    'that %r is correctly routed to separate databases.'
                 )
             else:
                 error_class, error_id = checks.Error, 'fields.E340'
@@ -1451,7 +1452,7 @@ class ManyToManyField(RelatedField):
                     "The field's intermediary table '%s' clashes with the "
                     "table name of '%s'." % (m2m_db_table, clashing_obj),
                     obj=self,
-                    hint=(error_hint % (m2m_db_table, clashing_obj)) if error_hint else None,
+                    hint=(error_hint % clashing_model) if error_hint else None,
                     id=error_id,
                 )
             ]
```

Comment (Qu4tro) [reply django/db/models/fields/related.py]: Thank you for the quick and comprehensive reply. Have updated the PR accordingly.
Will now correct the now outdated tests.

**Comment by Qu4tro** [comment]
Okay, all resolved. Have also squashed the commits and changed the commit message to reflect the changes in the PR title.

Let me know if there's any other changes required for this PR to move forward.

**Comment by felixxm** [comment]
@Qu4tro Thanks :+1: Welcome aboard :boat: 

I pushed minor edits to tests and updated the hint (_"Verify that **the table of** ..."_) to reuse `clashing_obj`.

**Comment by Qu4tro** [comment]
Thanks!

Yeah, I diffed with my local changes and it looks great. Much appreciated!

**Comment by felixxm** [comment]
Merged in 86908785076b2bbc31b908781da6b6ad1779b18b.
